### PR TITLE
Add OpenTracing support to inbound and outbound pipelines.

### PR DESCRIPTION
**Pull request by jmalloc**
This is approaching done-ness. I just wanted to push the PR now to get some feedback, specifically on the "event" names used in all of the pipeline stages.

As best I can tell, the only outstanding feature is logging application-logs (those made through the `MessageContext`) to the span, which should be fairly trivial.

### CODE REVIEW

File: src/ax/delayedmessage/tracing.go [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+import (
+	"time"
+
+	"github.com/jmalloc/ax/src/ax"
+	opentracing "github.com/opentracing/opentracing-go"
+	"github.com/opentracing/opentracing-go/log"
+)
+
+// traceIntercept adds a log a event representing the fact that a message has
```
Comment (koden-km) [inline src/ax/delayedmessage/tracing.go]:
```suggestion
// traceIntercept adds a log event representing the fact that a message has
```

File: src/ax/endpoint/tracing.go [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+	}
+
+	span.LogFields(
+		log.String("event", "transport.ack"),
+		log.String("message", "the message has been processed successfully"),
+	)
+}
+
+// traceInboundRetry adds a log event representing the time at which an inbound
+// message was was returned to the queue to be retried.
```
Comment (koden-km) [inline src/ax/endpoint/tracing.go]:
```suggestion
// message was returned to the queue to be retried.
```

File: src/ax/endpoint/tracing.go [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+			log.String("event", "transport.retry"),
+			log.String("message", "scheduling message for a delayed retry"),
+			log.String("delay-for", d.String()),
+			log.String("delay-until", t.Format(time.RFC3339Nano)),
+		)
+	}
+}
+
+// traceInboundReject adds a log event representing the time at which an inbound
+// message was was rejected due to the retry policy.
```
Comment (koden-km) [inline src/ax/endpoint/tracing.go]:
```suggestion
// message was rejected due to the retry policy.
```

**Comment by jmalloc** [comment]
Ah, I am also yet to add tracing through the saga subsystem.

**Comment by jmalloc** [comment]
This PR is now ready for a second round of review guys. All of the saga logging and appication-layer logging has been added now.

File: src/internal/tracing/log.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+		log.String("event", "error"),
+		log.String("message", err.Error()),
+		TypeName("error.kind", err),
+	)
+
+	ext.Error.Set(s, true)
+	s.LogFields(fields...)
+}
+
+// TypeName returns a log field contaning the name of v's type.
```
Comment (koden-km) [inline src/internal/tracing/log.go]:
```suggestion
// TypeName returns a log field containing the name of v's type.
```

File: src/internal/tracing/log.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+		),
+	)
+}
+
+// Duration returns a log field containing a human-readable representation of d.
+func Duration(k string, d time.Duration) log.Field {
+	return log.String(k, d.String())
+}
+
+// Time returns a log field containing a human-readable representation of t.
```
Comment (koden-km) [inline src/internal/tracing/log.go]: This comment mentions `t` but the function uses `d`. Rename the function to use `t`.

**Comment by danilvpetrov** [comment]
I can't see anything requiring changed.

File: src/internal/tracing/log.go (hunk 5 above)
Comment (jmalloc) [reply src/internal/tracing/log.go]: Fixed

**Pull request merged**